### PR TITLE
participant-integration-api: Switch to the v2 participant state API. [KVL-1002]

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
In which we [draw the rest of the owl](https://i.kym-cdn.com/photos/images/original/000/572/078/d6d.jpg).

Switching to the v2 participant state API means:

  - the API server and indexer expect v2 API traits
    - adapters are provided if you need to elevate your v1 API usage
  - the indexer internally uses v2 Updates
  - rejections are switched over to use the v2 format
  - Sandbox Classic uses v2 as the internal representation too (because it interacts directly with the underlying indexer representation, and is therefore tightly coupled)

_kvutils_ and other users of the `StandaloneApiServer` and `StandaloneIndexerServer` use the adapters.

### Changelog

- **[Integration Kit]** The API server and indexer have switched over to v2 of the participant-state API. You can continue to use the v1 API, but you will need to wrap your `ReadService` and `WriteService` objects in the `AdaptedV1ReadService` and `AdaptedV1WriteSerivce` classes.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/ConversionsSpec.scala
Comment (miklos-da) [inline ledger/participant-integration-api/src/test/suite/scala/platform/store/ConversionsSpec.scala:31]: Why is this error not a `RESOURCE_EXHAUSTED`?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/execution/LedgerTimeAwareCommandExecutor.scala
Comment (miklos-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/execution/LedgerTimeAwareCommandExecutor.scala:35]: You may want to introduce passing the ledger configuration in a separate PR, i.e., as we don't check configurations yet by the committer anyway (this would need removal of `SubmitterInfo.ledgerConfiguration` as well or always setting it to the default).

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/Update.scala
Comment (SamirTalwar-DA) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/Update.scala:234]: Not necessarily, but a `status` can have a lot more.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/DeduplicationPeriod.scala
Comment (SamirTalwar-DA) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/DeduplicationPeriod.scala:38]: Yes; this will be nested under `deduplicationPeriod`.

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/ConversionsSpec.scala
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/ConversionsSpec.scala:31]: I didn't want to change behavior. I agree that this is probably wrong.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/execution/LedgerTimeAwareCommandExecutor.scala
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/execution/LedgerTimeAwareCommandExecutor.scala:35]: I could, but it would be an ignored parameter and we'd have to suppress some warnings. I was trying to avoid this.

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/ConversionsSpec.scala
Comment (miklos-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/ConversionsSpec.scala:31]: Makes sense. Let's double-check the mapping between rejection reasons and gRPC error codes later.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/execution/LedgerTimeAwareCommandExecutor.scala
Comment (miklos-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/execution/LedgerTimeAwareCommandExecutor.scala:35]: No problem — after merging the PRs you've broken out from this one, passing the ledger configuration should be a minimal addition.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/Update.scala
Comment (miklos-da) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/Update.scala:234]: To put it in another way: I'm wondering if a single `RpcStatus` field would be enough for our needs?

Comment (SamirTalwar-DA) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/Update.scala:234]: Field, yes, but this is a trait, which needs to expose all these elements.

TBH I would like to remove the `status` here and in `FinalReason` but that's a big change I'd like to do separately.

File: ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/ledger/api/validation/SubmitRequestValidatorTest.scala
Comment (SamirTalwar-DA) [inline ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/ledger/api/validation/SubmitRequestValidatorTest.scala:165]:
```suggestion
```

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/SubmitterInfo.scala
Comment (SamirTalwar-DA) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/SubmitterInfo.scala:26]: Computers are tough.

```suggestion
  *                         same rules as the participant:
```

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/CommandsValidator.scala
Comment (fabiotudone-da) [inline ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/CommandsValidator.scala:50]: Is that already planned? If so, I think we can remove this TODO. If not, then I'd rather have a more general comment or none at all.

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/validation/ErrorFactories.scala
Comment (fabiotudone-da) [inline ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/validation/ErrorFactories.scala:38]: I get these are not used

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (fabiotudone-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:589]: Could you expand the param names here and in similar occurrences?

**Review by fabiotudone-da — APPROVED** [review]
LGTM (but consider that my context is somewhat limited)

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/CommandsValidator.scala
Comment (SamirTalwar-DA) [reply ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/api/validation/CommandsValidator.scala:50]: Yes, it is. I'll remove it.

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/validation/ErrorFactories.scala
Comment (SamirTalwar-DA) [reply ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/validation/ErrorFactories.scala:38]: You are correct!

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/JdbcLedgerDao.scala:589]: I'm trying not to touch `JdbcLedgerDao` more than necessary because of inevitable conflicts due to its size. That said, I hit these lines anyway, so why not?

**Review by miklos-da — APPROVED** [review]

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala]: Done. It is now `SubmissionIdGenerator.Random`.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/LedgerConfigProvider.scala
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/LedgerConfigProvider.scala:167]: This will fail on canton:
https://github.com/DACH-NY/canton/blob/2334662676d013852bd07d9a90ab44ea56f0aed2/community/participant/src/main/scala/com/digitalasset/canton/participant/sync/CantonSyncService.scala#L323

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala:157]: Have you checked that these errors will not be reported by canton or corda?

Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiSubmissionService.scala:157]: These errors no longer exist in the v2 enumeration, so they can't be any more. We can re-introduce them if a ledger driver has problems without them, but I expect it'll be fine.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/LedgerConfigProvider.scala
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/LedgerConfigProvider.scala:168]: @mziolekda pointed out that removing this might cause issues with ledgers + drivers such as Canton that don't support submitting configuration. As these drivers have other ways of managing configuration, I'm going to change `LedgerConfiguration` to allow them to state that we should _never_ submit a new configuration, instead just waiting indefinitely for one to show up. @oliverse-da has agreed this approach will work for Canton.

I'll make this change separately.

**Review by mziolekda — APPROVED** [review]
Looks good!
That's a big PR indeed. Mostly boiler plate on the storage side though.

**Comment by SamirTalwar-DA** [comment]
Agreed, it's huge. If it helps, imagine how big it would have been if we did it in one PR [instead of 17](https://github.com/digital-asset/daml/pulls?q=is%3Apr+sort%3Aupdated-desc+KVL-1002+in%3Atitle).

**Pull request merged**
